Source/utils/utf8.hpp

```diff
@@ -29,14 +29,29 @@ inline char32_t ConsumeFirstUtf8CodePoint(string_view *input)
 	return result;
 }
 
+/**
+ * Returns true if this is a byte that potentially starts a valid UTF-8 sequence.
+ *
+ * Well-formed UTF-8 sequences are described in table 3-7 of S3.9 of the Unicode Standard
+ * see: https://www.unicode.org/versions/Unicode14.0.0/ch03.pdf#G7404
+ *
+ * This is not an inverse of IsTrailUtf8CodeUnit. Byte values C0-C1, F5-FF are not valid anywhere in a UTF-8 sequence.
+ */
+constexpr bool IsLeadUtf8CodeUnit(unsigned char x)
+{
+	return x <= 0x7F || (x >= 0xC2 && x <= 0xF4);
+}
+
 /**
  * Returns true if this is a trailing byte in a UTF-8 code point encoding.
  *
- * A trailing byte is any byte that is not the heading byte.
+ * Trailing bytes all begin with 10 as the most significant bits, meaning they generally fall in the range 0x80 to
+ * 0xBF. Please note that certain 3 and 4 byte sequences use a narrower range for the second byte, this function is
+ * not intended to guarantee the character is valid within the sequence (or that the sequence is well-formed).
  */
 inline bool IsTrailUtf8CodeUnit(char x)
 {
-	return static_cast<signed char>(x) < -0x40;
+	return (x & 0xC0) == 0x80;

```

Comment:
The previous implementation did 1 comparison while this does an `&` and then a comparison. Any reason for this change?

https://godbolt.org/z/fTnGdW9jr

Reply:
This was mainly to express the criteria in a straightforward way, I was hoping that compilers would be able to optimise them to effectively the same instructions given the types involved :|

Oddly ARM as a target uses a different sequence for each version but ends up at the same instruction count. PowerPC uses fewer instructions for a bitmask test 😂 

I'll revert this with a comment explaining the reason

Reply:
So while playing around with this I realised this hits implementation defined behaviour on platforms where char is unsigned char by default due to unsigned to signed conversion not being strictly defined before C++20. I still reverted it but added a comment calling out the potential pitfall.

Or I think it can be avoided by casting x to `int` and comparing to a sign-extended int literal?

Reply:
getting into the weeds a bit but something like the following could also be used to avoid implementation defined behaviour :D

```cpp
#if defined(__cplusplus) && (__cplusplus >= 202002L)
	// The following is equivalent to a bitmask test (x & 0xC0) == 0x80
	// On x86_64 architectures it ends up being one instruction shorter
	// This would invokes implementation defined behaviour on platforms where the underlying type of char is unsigned
	// char which don't support C++20. From C++20 onwards unsigned to signed conversion is well defined
	return static_cast<signed char>(x) < static_cast<signed char>('\xC0');
#else
	return (x & 0xC0) == 0x80;
#endif
```

Reply:
I don't think we need to worry about it. The reason it became standard in C++20 is because almost everything (certainly everything DevilutionX can run on) is twos complement anyway: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0907r4.html#survey